cranelift/codegen/meta/src/shared/instructions.rs

```diff
@@ -4095,7 +4095,34 @@ pub(crate) fn define(
         Inst::new(
             "uwiden_high",
             r#"
-        Widen the high lanes of `x` using unsigned extension.
+        Lane-wise integer extended pairwise addition producing extended results
+        (twice wider results than the input)
+            "#,
+            &formats.unary,
+        )
+        .operands_in(vec![x])
+        .operands_out(vec![a]),
+    );
+
+    ig.push(
+        Inst::new(
+            "extended_pairwise_add_signed",
```

Comment:
An alternative approach is to introduce a binary pairwise addition IR operation, e.g. `iadd_pairwise`, which sounds like a more useful primitive operation - it could be used both by other existing Wasm instructions like `i32x4.dot_i16x8_s` and possible future additions (some of which have already been discussed - see [here](https://github.com/WebAssembly/simd/pull/382) for an example); in that case `extended_pairwise_add_signed` would become `iadd_pairwise(swiden_low, swiden_high)`. We need to do some pattern-matching for optimal AArch64 code generation (which is a single instruction), but I am not sure about other architectures - any thoughts from the people working on the backends?

cc  @abrown @cfallin @uweigand

Reply:
A general comment: in the past we were implicitly following the "make a new CLIF instruction for each Wasm SIMD instruction" paradigm so a lot of existing CLIF instructions could be split up in the same way you describe above. I think this is a thing that should be done (it just seems to make sense to simplify CLIF even at the risk of a few extra cycles during lowering) but I don't feel very strongly about it so I would be fine with either:
 - we do the pattern matching now AND we start a review of the existing SIMD instructions in a separate issue to clean up CLIF
 - we just start the review of existing SIMD instructions now and eventually remove the unnecessary CLIF instructions and switch to more pattern matching; this would mean following the "make a new CLIF instruction for each Wasm SIMD instruction" for the last few outstanding instructions

In summary, I don't mind if we do `iadd_pairwise` now or later, but I do think we need to commit to doing the same to the rest of SIMD CLIF.

Reply:
I forgot to link to [the discussion](https://github.com/bytecodealliance/wasmtime/pull/2982#discussion_r651773002) in #2982 for context.

@abrown While I don't have any strong feelings either, in #3044 I have already started with the first of your suggested approaches. In fact, the only Wasm SIMD operations that are not already covered by merged code or pull requests are the family of `i*.extmul_*_i*` operations, for which [the specification text](https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#extended-integer-multiplication) itself provides a simple way to express them in terms of other basic operations.

I agree that opening an issue for CLIF clean-up is definitely one of the steps forward, no matter what we decide to do.

Reply:
Opened https://github.com/bytecodealliance/wasmtime/issues/3045 for this.

Reply:
Just commented over in #3045 with thoughts on the general question; for this specific case I think it does seem reasonable to split into the pairwise add with extends on the inputs. But let's see what we come to over there!

Reply:
@akirilov-arm @cfallin Hi .. so the suggestion here is to remove the new IR instructions **extended_pairwise_add_signed** and **extended_pairwise_add_unsigned** and instead create one called **iadd_pairwise**. When lowering .. we would keep the lowering code that is there now, but pattern match for the instruction and type during lowering in order to get the input, or is the suggestion to scrape that lowering code too, actually lower swiden_low and swiden_high where we then are using the output of those particular instructions to be input to **iadd_pairwise** which needs to then be a new implementation from what is there now, essentially creating a new patch? I look at the change made in #2982 and it really is just avoiding creating another IR instruction while maybe adding a little logic of doing the matching, but keeps the original sequence. Here I am not sure if the suggestion is to go even further?

Reply:
IMHO the handling should be equivalent to the one for the `fcvt_from_uint` operation in #2982 - that is, there should be a code path that handles `iadd_pairwise` irrespective of its inputs, to serve as a fallback. Then right before that path there should be a check if the inputs match the pattern; if they do, execution should proceed through your current lowering code for `extended_pairwise_add_*`.

BTW @sparker-arm already has the common `iadd_pairwise` implementation and the AArch64-specific bits, but he is waiting for this pull request to get merged, so that you do not step on each other's toes. In the meantime, he may be able to offer some advice.

Reply:
Yeah, I don't mind posting this up if people are interested.

Reply:
Hi @spaker-arm .. can you post here?@akirilov-arm I still am not sure I understand what changes are in mind. Currently in code_translators there is this new instruction extended_pairwise_add_signed called here:

>         Operator::I16x8ExtAddPairwiseI8x16S => {
>             let a = pop1_with_bitcast(state, I8X16, builder);
>             state.push1(builder.ins().extended_pairwise_add_signed(a))
>         }
 
and new instruction extended_pairwise_add_unsigned called here:

>         Operator::I16x8ExtAddPairwiseI8x16U => {
>             let a = pop1_with_bitcast(state, I8X16, builder);
>             state.push1(builder.ins().extended_pairwise_add_unsigned(a))
>         }

You are saying to instead just call one instruction called iadd_pairwise and then lower from there? I think this is what you are saying and I'll go ahead and repush this.

Reply:
Sure, I'll post the whole thing up later today. But just to point out that we'll know the signedness because of the [u|s]widen inputs, for example:
```
        Operator::I16x8ExtAddPairwiseI8x16S => {
            let a = pop1_with_bitcast(state, I8X16, builder);
            let widen_low = builder.ins().swiden_low(a);
            let widen_high = builder.ins().swiden_high(a);
            state.push1(builder.ins().iadd_pairwise(widen_low, widen_high));
        }
```
This creates a bit more work in the AArch64 backend, but it does also mean that we can fallback to an ADDP when we don't successfully match the extending inputs. As a side queston, unless I'm mistaken wasm doesn't have a horizontal add - does anyone know why?